extension/flat_tensor/serialize/serialize.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <executorch/extension/flat_tensor/serialize/serialize.h>
+#include <executorch/extension/flat_tensor/serialize/schema_generated.h>
+#include <executorch/extension/flat_tensor/serialize/scalar_type_generated.h>
+
+#include <string>
+#include <fstream>
+
+namespace executorch {
+namespace extension {
+namespace flat_tensor {
+
+namespace {
+  size_t padding_required(size_t offset, size_t alignment) {
+    // Returns the padding required to align `offset` to `alignment`.
+    size_t remainder = offset % alignment;
+    if (remainder != 0) {
+        return alignment - remainder;
+    }
+    return 0;
+  }
+
+  size_t aligned_size(size_t input_size, size_t alignment) {
+    // Returns input_size padded up to the next whole multiple of alignment.
+    return input_size + padding_required(input_size, alignment);
+  }
+
+  void serialize_nulls(std::ostream& out, size_t num_bytes) {
+    for (size_t i = 0; i < num_bytes; i++) {
+      out.write("\0", 1);
+    }
+  }
+}
+
+ET_EXPERIMENTAL runtime::Error save_ptd(
+  const std::string& path,
+  const std::map<std::string, exec_aten::Tensor>& tensor_map,
+  const size_t tensor_alignment
+) {
+  // Assert the system is little endian
+  int n = 1;
+  if(*(char *)&n != 1) {
+    return runtime::Error::NotSupported;
+  }
+  // Create File
+  std::ofstream file;
+  file.open(path);
+  runtime::Error e = save_ptd(file, tensor_map, tensor_alignment);
+  file.close();
+  return e;
+}
+
+ET_EXPERIMENTAL runtime::Error save_ptd(
+  std::ostream& out,
+  const std::map<std::string, exec_aten::Tensor>& tensor_map,
+  const size_t tensor_alignment
+) {
+  // Assert the system is little endian
+  int n = 1;
+  if(*(char *)&n != 1) {
+    return runtime::Error::NotSupported;
+  }
+  // Create flatbuffer
+  flatbuffers::FlatBufferBuilder builder;
+
+  std::vector<flatbuffers::Offset<::flat_tensor::TensorMetadata>> tensors;
+  std::vector<flatbuffers::Offset<::flat_tensor::DataSegment>> buffers;
+
+  // Serialize the tensors.
+  size_t total_segment_size = 0;
+  for (const auto& [name, tensor] : tensor_map) {
+    auto name_offset = builder.CreateString(name);
+    // Serialize the tensor metadata.
+    auto tensor_metadata = ::flat_tensor::CreateTensorMetadata(
+      builder,
+      name_offset,
+      static_cast<executorch_flatbuffer::ScalarType>(tensor.scalar_type()),
+      builder.CreateVector(tensor.sizes().data(), tensor.sizes().size()),
+      builder.CreateVector(tensor.dim_order().data(), tensor.dim_order().size()),
+      0, // segment index
+      total_segment_size);
+    tensors.push_back(tensor_metadata);      // Precalculate the size of the data blob
+    total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
+  }
+  // Only have one segment
```

Comment:
Why is there only one segment? Seems like we'd want to be able to load and free large tensors independently

Reply:
Thats the behavior of the python serializer today. I was just mimicking it. Theres not an immediate need to separate today, it will come if we start combining mutable and non mutable weights in the same .ptd for training, or with delegates presumably. Can add that functionality later though with no bc breaks.